contracts/IndexTemplate.sol

```diff
@@ -328,114 +337,129 @@ contract IndexTemplate is InsureDAOERC20, IIndexTemplate, IUniversalMarket {
-                //Target credit allocation for a pool
-                uint256 _target = (_targetCredit * _allocation) /
-                    _allocatablePoints;
-                //get how much has been allocated for a pool and get how much liquidty is available to withdraw
-                uint256 _current;
-                uint256 _available;
-                (_current, _available) = IPoolTemplate(_pool).pairValues(address(this));
-                //if needed to withdraw credit but unable, then withdraw all available.
-                //Otherwise, skip.
+                uint256 _allocPoint = allocPoints[_pool];
+
+                //Get "current allocation on the pool" & "Pool's _availableBalance()"
+                (uint256 _currentCredit, uint256 _availableBalance) = IPoolTemplate(_pool).pairValues(address(this));
+
                 if(
+                    //Paying out pool
                     IPoolTemplate(_pool).marketStatus() == IPoolTemplate.MarketStatus.Payingout
                 ){
-                    totalAllocatedCredit -= _current;
-                    _poolList[i].addr = address(0);
-                    _allocatable = _safeMinus(_allocatable, _current);
-                    _allocatablePoints -= _allocation;
-                } else if (
-                    _current > _target && _current - _target > _available
-                ) {
-                    IPoolTemplate(_pool).withdrawCredit(_available);
-                    _totalAllocatedCredit -= _available;
-                    _poolList[i].addr = address(0);
-                    _allocatable -= _current - _available;
-                    _allocatablePoints -= _allocation;
+                    //Reserve a slot for all _currentCredit amount. 
+                    _allocatableCredit = _safeMinus(_allocatableCredit, _currentCredit); 
+
+                    _poolList[i].addr = address(0); //done for this pool
+                    _afterTotalAllocPoint -= _allocPoint;
+
                 } else if (
+                    //When pool is paused, withdraw all
                     IPoolTemplate(_pool).paused()
                 ) {
-                    if(_current > _available){
-                        IPoolTemplate(_pool).withdrawCredit(_available);
-                        totalAllocatedCredit -= _available;
-                        _poolList[i].addr = address(0);
-                        _allocatable -= _safeMinus(_allocatable, _available);
-                        _allocatablePoints -= _allocation;
+                    if(_currentCredit > _availableBalance){
+                        //withdraw as much as possible
+                        IPoolTemplate(_pool).withdrawCredit(_availableBalance);
+                        _totalAllocatedCredit -= _availableBalance;
+                        
+                        //Reserve a slot for the credits still in the pool
+                        _allocatableCredit -= _currentCredit - _availableBalance;
```

Comment:
Should use `_safeMinus()` to avoid underflow when `_allocatableCredit` is lower than `_currentCredit - _availableBalance`.